### PR TITLE
[Keyboard] Add TypeTheory Keyboard

### PR DESCRIPTION
**Pull request by mgvx**
## Description

Add support for the [TypeTheory Keyboard](https://typetheory.co.uk).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/typetheory/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/typetheory/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/typetheory/typetheory.c [hunk 2]
```diff
@@ -0,0 +1 @@
+#include "typetheory.h"
```
Comment (drashna) [inline keyboards/typetheory/typetheory.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/typetheory/rules.mk [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+LTO_ENABLE       = yes
+
+BOOTMAGIC_ENABLE = yes
+MOUSEKEY_ENABLE  = no
+EXTRAKEY_ENABLE  = yes
+CONSOLE_ENABLE   = yes
+COMMAND_ENABLE   = yes
+NKRO_ENABLE      = yes
+UNICODE_ENABLE   = no
+BLUETOOTH_ENABLE = no
```
Comment (fauxpark) [inline keyboards/typetheory/rules.mk:14]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
SPLIT_KEYBOARD = yes
LTO_ENABLE = yes
```

File: keyboards/typetheory/readme.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+1. Press the RESET key on the keyboard, this is RAISE+RIGHT SHIFT on the default layout.
+2. With the keyboard unplugged, hold SPACE + B while plugging in the keyboard.
+3. Open up the keyboard half, and press the reset button on the Puchi-C board
+
+Once reset, you can flash the left/right half by using one of these commands respectively:
+
+    make typetheory:default:dfu-split-left
+    make typetheory:default:dfu-split-right
+
+Tip: if desired, you can flash different layouts to the left/right side. Then, the entire keyboard will use the 'left' layout when the left side is plugged into the computer, and vice-versa for the right side. Do not attempt to plug in both keyboard halves into the computer at once.
```
Comment (fauxpark) [inline keyboards/typetheory/readme.md:29]:
```suggestion
# TypeTheory Keyboard

![Picture of TypeTheory Keyboard](https://typetheory.co.uk/thumb/front.jpg)

The [TypeTheory Keyboard](https://typetheory.co.uk/) is a split ergonomic keyboard inspired by the [Dactyl-Manuform](https://github.com/tshort/dactyl-keyboard) and the Kinesis Advantage2. It uses two Puchi-C microcontrollers, with physical reset buttons.

* Keyboard Maintainer: Marius Gavrilescu
* Hardware Supported: TypeTheory Keyboard
* Hardware Availability: https://typetheory.co.uk

Make example for this keyboard (after setting up your build environment):

    make typetheory:default

Flashing example for this keyboard:

    make typetheory:default:dfu-split-left
    make typetheory:default:dfu-split-right

See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).

## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the Puchi-C
* **Keycode in layout**: Press the key mapped to `RESET` (`RAISE`+Right Shift on the default keymap)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/typetheory/rules.mk [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = yes        # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
```
Comment (fauxpark) [inline keyboards/typetheory/rules.mk:21]:
```suggestion
```

File: keyboards/typetheory/info.json [hunk 6]
```diff
@@ -0,0 +1,76 @@
+{
+    "keyboard_name": "TypeTheory Keyboard",
+    "url": "https://typetheory.co.uk",
+    "maintainer": "mgvx",
+    "width": 17,
+    "height": 8,
```
Comment (fauxpark) [inline keyboards/typetheory/info.json:6]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Pull request closed without merging**